### ISSUE TITLE
feat(phase-2b): Statistical discovery engine

### ISSUE THREAD

**Issue by omniscient** [omniscient/markethawk#22]
## Goal

After Phase 2a has populated enriched features across a meaningful backtest dataset (target: 2 years × 10 signal types ≈ 5,000+ events), run statistical analysis to discover which features are predictive of subsequent returns at each outcome interval (1h, 4h, EOD, 1d, 2d, 5d).

## Context

Part of the [Phase 2 signal ranking initiative](#PARENT). This is the **discovery phase** — the output is not a deployed model but a set of findings: which features matter, how much, and under what conditions.

---

### 1. Correlation analysis
For each feature in `indicators`, compute Pearson and Spearman correlation against `pct_change` at each outcome interval. Surfaced as a heatmap in EdgeExplorer.

### 2. Feature importance
Fit a simple LightGBM regressor (target = 1h/4h/1d return) on the enriched feature set. Use SHAP values to rank feature importance. Non-deployed — run as a Celery task on demand.

### 3. Signal cluster analysis
K-means or DBSCAN on the feature vector to discover natural signal archetypes. Label each cluster with its median return profile. Answer: "do high-return events share a profile?"

### 4. Conditional statistics
For each feature × interval combination, compute:
- Median return
- Win rate (return > 0)
- Sharpe ratio
- Sample size

Example finding: *"Signals with `volume_anomaly_score` > 3.0 AND `market_context = risk_on` have median +2.1% return by 4h, n=147"*

---

## Output artifacts

1. **EdgeExplorer panel**: Feature correlation heatmap, filterable by signal type and interval
2. **Cluster report**: Named archetypes with return profiles (e.g. "Morning gap + risk-on = cluster A")
3. **Feature weight table**: Input for Phase 2c ranker
4. **Celery task**: `analyze_signal_features()` — runs on demand or nightly after backtest

---

## Dependencies
- Phase 2a must be complete and backtest data populated
- Minimum dataset: ~500 events per signal type for meaningful statistics

---

## Acceptance criteria
- [ ] `analyze_signal_features` Celery task runs without error on full dataset
- [ ] Correlation heatmap visible in EdgeExplorer
- [ ] Cluster analysis output stored in DB (new `signal_cluster` column or table)
- [ ] Feature weight table exportable as JSON (input for Phase 2c)

**Comment by omniscient** [comment]
## Refinement Pipeline — Spec Generated

**Spec:** [Docs/superpowers/specs/2026-05-14-statistical-discovery-engine-design.md](https://github.com/omniscient/markethawk/blob/refine/issue-22-feat-phase-2b---statistical-discovery-en/Docs/superpowers/specs/2026-05-14-statistical-discovery-engine-design.md)
**Branch:** [`refine/issue-22-feat-phase-2b---statistical-discovery-en`](https://github.com/omniscient/markethawk/tree/refine/issue-22-feat-phase-2b---statistical-discovery-en)

### Summary

The spec defines a statistical discovery pipeline (Phase 2b) that turns the Phase 2a backtest dataset into actionable findings for the Phase 2c ranker. A new `analyze_signal_features()` Celery task joins `ScannerEvent.indicators` with `ScannerOutcomeSnapshot.pct_change` per interval, computes Pearson/Spearman correlations, fits LightGBM + SHAP for feature importances, and runs K-means clustering to produce named signal archetypes. Results are stored in two new tables (`SignalAnalysisRun`, `SignalCluster`) and surfaced via three new API endpoints and a new correlation heatmap section in EdgeExplorer.

### Brainstorming Q&A

> **Q:** Which outcome model should `analyze_signal_features()` use as the return target — `ScannerOutcomeSnapshot` (per-interval pct_change) or `ScannerOutcomeSummary` (aggregates like mfe_pct, r_multiple)?
> **A:** Use `ScannerOutcomeSnapshot` with `status = 'captured'`. The analysis is explicitly framed as "pct_change at each outcome interval" — that maps directly to `ScannerOutcomeSnapshot.interval_key` + `pct_change`. Use `ScannerOutcomeSummary.is_complete = True` as an efficiency gate to scope which event IDs enter analysis (ensures all six snapshot intervals are captured before analysing an event).

> **Q:** Are `lightgbm`, `shap`, or `scikit-learn` already in `requirements.txt`? Should we add LightGBM + SHAP or substitute scikit-learn?
> **A:** None are currently installed — all would be new additions. Add LightGBM + SHAP as specified in the issue. Since neither scikit-learn nor lightgbm is already present, there is no "free" lighter path. SHAP TreeExplainer produces per-feature importance values directly; RandomForest's impurity-based `feature_importances_` would be a meaningful capability downgrade. (`scikit-learn` and `scipy` are also needed for KMeans and Spearman respectively, so the full new dependency list is: `lightgbm`, `shap`, `scikit-learn`, `scipy`.)

> **Q:** For cluster storage, should we add columns to `ScannerEvent` or create a separate `SignalCluster` table?
> **A:** Use a separate `SignalCluster` table with a nullable FK on `ScannerEvent`. The codebase pattern (e.g. `ScannerOutcomeSummary`, `UniverseQualityReport`) consistently normalises derived entity data into separate tables rather than denormalising into the parent. Cluster archetype metadata (centroid, return profile) is shared across many events and belongs on its own entity. The spec uses `analysis_run_id → signal_analysis_runs.id` (a new anchor table) rather than `run_id → scanner_runs.id` because analysis runs are a distinct concept from scan executions.

> **Q:** What trigger mechanism should `analyze_signal_features()` use for on-demand calls, and how should nightly scheduling work? What format for the feature weight table?
> **A:** On-demand: `POST /api/outcomes/analyze` returns `202 { "task_id": "..." }`, following the `POST /api/scanner/run` pattern in the existing outcomes router. Nightly: add to Celery Beat at `11:00 UTC weekdays` (6 AM ET / 7 AM EDT — after overnight outcome capture settles, before the pre-market scan window). Feature weights: `GET /api/outcomes/analysis/latest` returns the most recent `SignalAnalysisRun`'s results as JSON — no file download endpoint needed.

> **Q:** Should the correlation heatmap be a new endpoint or an extension of `/edge-stats`? Should it be a new page or a new section in EdgeExplorer? Which rendering approach for the heatmap?
> **A:** New endpoint `GET /api/outcomes/correlations` in the existing outcomes router (extending `/edge-stats` would break its contract). New card section at the bottom of the existing `EdgeExplorer.tsx` page — no separate route. CSS-colored grid/table for rendering: no new charting library needed. Color scale: negative = red `#EF4444`, zero = dark gray `#374151`, positive = green `#10B981`. Extract to `CorrelationHeatmap.tsx` component.

### Requirements

- `analyze_signal_features()` Celery task runs on demand and nightly (Beat at 11:00 UTC weekdays)
- Analyzes only `ScannerOutcomeSummary.is_complete = True` events; uses `ScannerOutcomeSnapshot.pct_change` per interval as target
- Pearson + Spearman correlation per feature × interval
- LightGBM + SHAP feature weights per interval
- K-means (k=6 default) cluster archetypes with auto-generated labels
- Conditional statistics (median return, win rate, Sharpe, n) per cluster × interval
- New tables: `SignalAnalysisRun`, `SignalCluster`; new FK column on `ScannerEvent`
- Three new endpoints: `POST /api/outcomes/analyze`, `GET /api/outcomes/correlations`, `GET /api/outcomes/analysis/latest`
- New `CorrelationHeatmap` component + section in `EdgeExplorer.tsx`
- Alembic migration; no new npm packages

### Approach

Pre-compute all statistical outputs inside the Celery task and store them in `SignalAnalysisRun` as JSONB. API endpoints serve stored results, keeping p50 response times well under 100ms regardless of dataset size. K-means with configurable k is used for clustering in preference to DBSCAN, which requires per-dataset epsilon tuning incompatible with automated nightly runs.

### Assumptions

- Phase 2a is complete: `ScannerOutcomeSnapshot` / `ScannerOutcomeSummary` tables exist and are populated
- At least 500 events with `is_complete = True` must exist before the task will produce results (returns an error otherwise)
- `ScannerEvent.indicators` keys are consistent enough across scanner types to build a shared feature matrix; string/boolean values are coerced to float or dropped

### Next Steps

- ✅ **Approve spec** — move the issue to the **Refined** column on the project board. The scheduler will automatically trigger plan generation.
- ✏️ **Request changes** — leave a comment on this issue with your feedback, then re-run:
  ```bash
  docker compose --profile factory run --rm dark-factory "Refine issue #22"
  ```
- ❓ **Need to discuss** — add the `needs-discussion` label to pause automation.

---
*Posted by MarketHawk Refinement Pipeline*

**Comment by omniscient** [comment]
📋 **Refinement Pipeline** — Starting plan generation and architect validation.

---
*Posted by MarketHawk Backlog Scheduler*

**Comment by omniscient** [comment]
## Refinement Pipeline — Plan Generated

**Plan:** [Docs/superpowers/plans/2026-05-14-statistical-discovery-engine.md](https://github.com/omniscient/markethawk/blob/main/Docs/superpowers/plans/2026-05-14-statistical-discovery-engine.md)
**Branch:** [`main`](https://github.com/omniscient/markethawk/tree/main)
**Tasks:** 8 tasks, 41 steps

### Task Overview

1. **Add Python dependencies** — Append `lightgbm==4.6.0`, `shap==0.47.2`, `scikit-learn==1.6.1`, `scipy==1.15.3` to `requirements.txt` and verify install in Docker
2. **Create SignalAnalysisRun and SignalCluster models + migration** — Two new ORM models, add `signal_cluster_id` FK to `ScannerEvent`, register in `__init__.py`, generate and apply Alembic migration with correct FK ordering
3. **Create StatisticalDiscoveryService (TDD)** — Pure-Python service with 6 methods (`build_feature_matrix`, `compute_correlations`, `compute_shap_weights`, `run_kmeans`, `compute_conditional_stats`, `generate_label`); 12 unit tests written first
4. **Add analyze_signal_features Celery task + Beat schedule** — Celery task with full SessionLocal scaffolding, queries complete events, runs all service methods, persists clusters, back-fills `ScannerEvent.signal_cluster_id`; Beat entry at 11:00 UTC weekdays
5. **Add analysis schemas** — Pydantic response schemas (`CorrelationResponse`, `LatestAnalysisResponse`, `ClusterSummary`, etc.) with `ConfigDict(from_attributes=True)` on response models
6. **Add API endpoints (TDD)** — Three endpoints on `outcomes.py` router (`POST /analyze`, `GET /correlations`, `GET /analysis/latest`); 6 integration tests + fixture written first; curl smoke-tested
7. **Create CorrelationHeatmap frontend component (TDD)** — TypeScript API layer (`analysis.ts`) + React component with Pearson/Spearman toggle and CSS color interpolation; verified with `tsc --noEmit`
8. **Add Feature Correlations card to EdgeExplorer** — `useQuery` + `useMutation` hooks, new `<Card>` section with heatmap, Run Analysis button, loading/empty states; verified with `tsc --noEmit`

### Architect Review

> **Cycle 1:**
> **Verdict:** Issues Found
> **Feedback:**
> - [Task 6 / POST endpoint]: Spec showed `AsyncSession` but plan should use synchronous `Session = Depends(get_db)` — the existing outcomes.py router is synchronous
> - [Task 8 / Toast]: No toast library in codebase; spec should prescribe inline `alert()` feedback instead
> - [Task 3 / generate_label]: Missing from service outline but called in Task 4 — needs explicit definition in Task 3
> - [Task 4 / Celery scaffolding]: Missing `SessionLocal()` / `try/finally db.close()` scaffolding and bulk-update code block
> - [Task 2 / Migration ordering]: Under-specified — no concrete code showing correct `op.create_table` ordering
> - [Task 5 / model_config]: `ConfigDict(from_attributes=True)` missing from response schemas
>
> **Changes made:** Confirmed plan already has synchronous `Session`, `alert()`, `generate_label`, and full Celery scaffolding (architect was reviewing an abbreviated summary). Fixed: added `ConfigDict(from_attributes=True)` to `CorrelationResponse`, `ClusterSummary`, and `LatestAnalysisResponse` in Task 5; added concrete migration ordering code snippet in Task 2.

> **Cycle 2:**
> **Verdict:** Approved
> **Feedback:**
> - Advisory: `test_trigger_analysis_returns_202` mock target — patch `app.tasks.analyze_signal_features` works correctly for local imports (patches at source, picked up at import time); no change needed
> - Advisory: Check `backend/tests/fixtures/__init__.py` and add import if needed
> - Advisory: Task 8 JSX insertion point should be inside the conditional rendered branch (after last existing Card, before closing `</>`)
> - Advisory: Consider adding `test_trigger_analysis_with_scanner_type` test
>
> **Changes made:** None — plan is correct on all flagged points.

### Next Steps

- ✅ **Approve plan** — move the issue to the **Ready** column on the project board. The scheduler will automatically start implementation.
- ✏️ **Request changes** — leave a comment on this issue with your feedback, then re-run:
  ```bash
  docker compose --profile factory run --rm dark-factory "Plan issue #22"
  ```
- ❓ **Need to discuss** — add the `needs-discussion` label to pause automation.

---
*Posted by MarketHawk Refinement Pipeline*

**Comment by omniscient** [comment]
It looks good to me, let's implement this

**Comment by omniscient** [comment]
## Dark Factory Run — New implementation

**PR:** #57

**Branch:** `feat/issue-22-feat-phase-2b---statistical-discovery-en`

### Changes

# Phase 2b Implementation — Statistical Discovery Engine (Issue #22)

## Files Created

### Backend
- `backend/app/models/signal_analysis_run.py` — `SignalAnalysisRun` ORM model (anchor table)
- `backend/app/models/signal_cluster.py` — `SignalCluster` ORM model
- `backend/app/alembic/versions/7ba47256a679_add_signal_analysis_tables.py` — migration (creates both tables + signal_cluster_id FK on scanner_events)
- `backend/app/services/statistical_discovery.py` — `StatisticalDiscoveryService` (6 pure-Python methods)
- `backend/app/schemas/analysis.py` — Pydantic response schemas with ConfigDict(from_attributes=True)
- `backend/tests/services/test_statistical_discovery.py` — 12 unit tests (all pass)
- `backend/tests/fixtures/analysis.py` — DB seed helper for integration tests
- `backend/tests/api/test_analysis.py` — 6 integration tests (all pass)

### Frontend
- `frontend/src/api/analysis.ts` — API layer (fetchCorrelations, fetchLatestAnalysis, triggerAnalysis)
- `frontend/src/components/CorrelationHeatmap.tsx` — CSS-colored correlation table with Pearson/Spearman toggle

## Files Modified

### Backend
- `backend/requirements.txt` — added lightgbm==4.6.0, shap==0.47.2, scikit-learn==1.6.1, scipy==1.15.3
- `backend/app/models/__init__.py` — registered SignalAnalysisRun, SignalCluster
- `backend/app/models/scanner_event.py` — added signal_cluster_id FK column
- `backend/app/tasks.py` — added analyze_signal_features Celery task
- `backend/app/core/celery_app.py` — added nightly Beat entry (11:00 UTC weekdays)
- `backend/app/routers/outcomes.py` — added 3 new endpoints (POST /analyze, GET /correlations, GET /analysis/latest)

### Frontend
- `frontend/src/pages/EdgeExplorer.tsx` — added useMutation hook, analysis hooks, Feature Correlations card

### Documentation
- `ARCHITECTURE.md` — added statistical_discovery.py service, outcomes.py router, SignalAnalysisRun/SignalCluster models, EdgeExplorer page update, analyze_signal_features task
- `PROJECT_STRUCTURE.md` — added entries for all new/modified files

## Tests Added
- 12 unit tests for `StatisticalDiscoveryService` (build_feature_matrix, compute_correlations, compute_shap_weights, run_kmeans, compute_conditional_stats, generate_label)
- 6 integration tests for the 3 new API endpoints

## Migrations Created
- `7ba47256a679_add_signal_analysis_tables.py` — creates `signal_analysis_runs`, `signal_clusters`, adds `scanner_events.signal_cluster_id` FK (manually trimmed orphan-table drops from autogenerate output)

## Decisions / Trade-offs
- Orphan tables (alert_history, market_data_cache, scanner_execution_log, user_preferences, alert_configs) detected by Alembic autogenerate were removed from the migration to avoid accidental data loss — they have no ORM models but exist in the DB.
- The dark factory container accesses the git repo at `/workspace/markethawk` while the application Docker containers mount from the Windows host path — files were synced to Docker via `docker cp` for runtime verification.
- Local Python deps (lightgbm, shap, sklearn, scipy) were installed in the dark factory container alongside the Docker container install for running unit tests locally via testcontainers.
- Final test run: 278 passed, 0 failed. TypeScript: 0 errors.

### Preview Environment

| Service | URL |
|---------|-----|
| Frontend | http://localhost:12233 |
| Backend API | http://localhost:12280 |
| API Docs | http://localhost:12280/docs |
| PostgreSQL | `localhost:12254` |
| Redis | `localhost:12263` |

### Commands
```bash
# Iterate after review
docker compose --profile factory run --rm dark-factory "Continue issue #22"

# Tear down when done
docker compose --profile factory run --rm dark-factory "Close issue #22"
```

---
*Posted by MarketHawk Dark Factory*

**Comment by omniscient** [comment]
Perfect, merge and close

**Comment by omniscient** [comment]
Merging PR and tearing down preview environment.

**Comment by omniscient** [comment]
Dark factory: preview torn down, PR #57 merged. Issue complete.

**Comment by omniscient** [comment]
Merging PR and tearing down preview environment.
